### ISSUE TITLE
HTTP and TCP timeout configuration can't working separately

### ISSUE THREAD

**Issue by hubv** [istio/istio#35418]
### Bug Description

We tested two Scenarios to verify if HTTP and TCP timeout configuration in Istio can be working separately.

<html>
<body>
<div style="box-sizing: border-box; font-family: &quot;Segoe UI&quot;, system-ui, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px;">

<div class="copy-paste-block">

  | HTTP layer | TCP layer
-- | -- | --
Timeout | Timeout in HTTPRoute in VirtualService | connectTimeout in ConnectionPoolSettings in DestinationRule

</div></div>
</body>
</html>

**Scenario1: For both destinations set TCP timeout= 6s, HTTP timeout=2s.**

**Configuration:**
configured 2 destinations in VirtualService. D1 and  D2

**Observations:**

1. Request sent at 18:00:09
2. TCP re-transmission happened twice to D1 i.e at18:00:10 and 18:00:12 (as per config TCP re-transmission looks ok)
3. HTTP timeout applied to D2 after 2s i.e at 18:00:11 (as per config, HTTP timeout value is lesser value compared to TCP timeout, so Envoy decided to look for alternative destination  i.e D2
4. HTTP timeout applied to D2 after 2s and Envoy sent back 504 gateway timeout to downstream at 18:00:13 (As per HTTP timeout config this looks fine)

**Scenario1 Conclusion:**

istio will consider timeout value whichever is lesser. If both TCP and HTTP timeout values are set, they don't wait for each others decision.

  

**Scenario2: Destination1: set TCP timeout= 6s, HTTP timeout=9s. Destination2: et TCP timeout= 6s, HTTP timeout=2s**

**Configuration:**

configured 2 destinations in VirtualService. D1 and  D2. 

**Observations:**

1. Request sent at 18:14:13 (~18:14:14)
2. TCP re-transmission happened twice to D1 i.e at 18:14:15 and 18:14:17 (as per config TCP re-transmission looks fine)
3. TCP timeout applied to D1 after around 6s at 18:14:20 (as per config, TCP timeout value is lesser value compared to HTTP timeout, so Envoy decided to look for alternative destination i.e D2)
4. HTTP timeout applied to D2 after 2s and Envoy sent back 504 gateway timeout to downstream at 18:14:22 (As per HTTP timeout config this looks fine)

 

**Scenario2 Conclusion:**

istio will consider timeout value whichever is lesser. If both TCP and HTTP timeout values are set, they don't wait for each others decision

 

**Final Conclusion:**

Each parameter cannot be worked independently. If lesser timeout value is considered first.
Is it expected behaviour based on the specification? If not, is it bug?
How to set TCP and HTTP timeout separately in Istio?

Regards

### Version

```prose
[root@vincent-control-01 ~]# kubectl version --short
Client Version: v1.18.8
Server Version: v1.18.8
[root@vincent-control-01 ~]# 2108/istio-1.10.3/bin/istioctl version --short
client version: 1.10.3
control plane version: istio-release-1.10.4
data plane version: 1.10.4 (6 proxies)
```

**Comment by pangsq** [comment]
> HTTP timeout applied to D2 after 2s i.e at 18:00:11 (as per config, HTTP timeout value is lesser value compared to TCP timeout, so Envoy decided to look for alternative destination i.e D2

I think the reason is not right. It should be that Envoy gives a retry here because of `connect-failure`, according to [retry.DefaultPolicy](https://github.com/istio/istio/blob/fa911b97449b5f793deabd964f9bfa89e48b9ad1/pilot/pkg/networking/core/v1alpha3/route/retry/retry.go#L36)

Once the tcp `timeout` or http `timeout` triggered, the request is failed. Whether envoy continues to try the request or not depends on `retry policy`.

The timeline how envoy processs the request in my understanding is ：
```
# if envoy has not built the connection to upstream instance
envoy received entire request from downstream(A)
-> envoy begin to build connection to upstream instance(B)
-> connection built successfully(C)
-> envoy begin to send request on the connection(D)
-> ...(some steps ignored) 
-> envoy received complete response from upstream instance(H)
```
HTTP `timeout` calculates `A -> H` ,  TCP `timeout`(which is actually `connect time`) calculates `B -> C`. If the connection is already built before, `B/C` are not need in this request.

In conclusion, HTTP and TCP `timeout` do work separately.

More info about `timeout` in istio/envoy is described here: [timeouts](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts)

**Comment by hubv** [comment]
HI @pangsq,

Any update?

Regards,
Vincent
